### ISSUE TITLE
[Phase 5] Deep Learning Models

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/ecg-arrhythmia-classification#6]
## Phase 5: Deep Learning Models

**Parent:** #1
**Depends on:** #3

### Objectives
Implement and train deep learning models (1D CNN, LSTM with Attention) on raw ECG beats.

### Tasks

- [ ] Implement 1D CNN architecture
- [ ] Implement Bidirectional LSTM with Attention
- [ ] Create PyTorch Dataset and DataLoader
- [ ] Implement training loop with validation
- [ ] Add learning rate scheduling (ReduceLROnPlateau)
- [ ] Add early stopping
- [ ] Implement class-weighted loss
- [ ] Save best model checkpoints
- [ ] Add TensorBoard logging (optional)

### Files to Create/Modify

| File | Action | Description |
|------|--------|-------------|
| `src/models.py` | Update | Add CNN and LSTM classes |
| `src/train.py` | Update | Add PyTorch training loop |
| `src/dataset.py` | Create | PyTorch Dataset class |

### Code Reference

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class CNN1D(nn.Module):
    def __init__(self, num_classes=5):
        super().__init__()
        self.conv1 = nn.Conv1d(1, 32, kernel_size=7, padding=3)
        self.bn1 = nn.BatchNorm1d(32)
        self.conv2 = nn.Conv1d(32, 64, kernel_size=5, padding=2)
        self.bn2 = nn.BatchNorm1d(64)
        self.conv3 = nn.Conv1d(64, 128, kernel_size=3, padding=1)
        self.bn3 = nn.BatchNorm1d(128)
        self.conv4 = nn.Conv1d(128, 256, kernel_size=3, padding=1)
        self.bn4 = nn.BatchNorm1d(256)

        self.pool = nn.MaxPool1d(2)
        self.dropout = nn.Dropout(0.3)
        self.fc1 = nn.Linear(256, 128)
        self.fc2 = nn.Linear(128, num_classes)

    def forward(self, x):
        # x: (batch, 1, 360)
        x = self.pool(F.relu(self.bn1(self.conv1(x))))
        x = self.dropout(x)
        x = self.pool(F.relu(self.bn2(self.conv2(x))))
        x = self.dropout(x)
        x = self.pool(F.relu(self.bn3(self.conv3(x))))
        x = self.dropout(x)
        x = self.pool(F.relu(self.bn4(self.conv4(x))))
        x = x.mean(dim=-1)  # Global average pooling
        x = F.relu(self.fc1(x))
        x = self.dropout(x)
        return self.fc2(x)


class LSTMAttention(nn.Module):
    def __init__(self, input_size=1, hidden_size=128, num_classes=5):
        super().__init__()
        self.lstm = nn.LSTM(input_size, hidden_size, num_layers=2,
                           bidirectional=True, batch_first=True, dropout=0.2)
        self.attention = nn.Linear(hidden_size * 2, 1)
        self.fc = nn.Linear(hidden_size * 2, num_classes)

    def forward(self, x):
        # x: (batch, 360, 1)
        lstm_out, _ = self.lstm(x)
        attn_weights = F.softmax(self.attention(lstm_out), dim=1)
        context = (lstm_out * attn_weights).sum(dim=1)
        return self.fc(context)
```

### Training Configuration

```python
config = {
    'batch_size': 64,
    'learning_rate': 0.001,
    'epochs': 50,
    'patience': 10,  # Early stopping
    'lr_patience': 5,  # LR scheduler
    'lr_factor': 0.5,
}
```

### Definition of Done

- [ ] CNN trains and converges (loss decreases)
- [ ] LSTM trains and converges
- [ ] Validation accuracy >95%
- [ ] Early stopping triggers appropriately
- [ ] Best model checkpoint saved
- [ ] No CUDA OOM errors (adjust batch size if needed)

### Expected Results

| Model | Accuracy | F1 (weighted) |
|-------|----------|---------------|
| 1D CNN | 97-99% | 0.96-0.98 |
| LSTM+Attention | 96-98% | 0.95-0.97 |

### Technical Notes

**For junior developers:**
- Reshape data: CNN expects (batch, channels, length), LSTM expects (batch, length, features)
- Use `torch.cuda.is_available()` to check GPU
- Start with smaller batch size (32) if memory issues
- BatchNorm helps stabilize training
- Attention weights can be visualized for interpretability
